Packet++/header/Packet.h

```diff
@@ -330,6 +333,40 @@ namespace pcpp
 		return (TLayer*)curLayer;
 	}
 
+
+	/**
+	 * @class PacketDecoder
+	 * A class for parsing a packet. It provides the ability to parse the next layers for partially decoded packets
+	 *
+	 * For example: the gradually parsing the RADIUS and HTTP protocols with the next logic:
+	 *  1. create the packet by parsing until IP layer, in order to pass the partially parsed packet into IPReassembly::processPacket.
+	 *     It gains the performance because the parsing of the next layers is useless for the fragmented IP datagrams
+	 *  2. parse the reassembled/non-fragmented packets until transport layer (UDP, TCP)
+	 *  3. pass the partially parsed TCP packets into TcpReassembly::ReassemblePacket to reconstruct a TCP stream before HTTP protocol parsing
+	 *  4. parse the next layers of UDP packets until RADIUS layer
+	 */
+	class PacketDecoder
+	{
+	public:
+
+		/**
+		 * A static method that attempts to parse next layer
+		 * @param[in] pkt The pointer to packet to be decoded
+		 * @return A pointer to the parsed layer or NULL if there are no more layers
+		 */
+		static Layer* parseNextLayer(Packet* pkt);
+
+		/**
+		 * A method decodes the next layers in a packet
+		 * @param[in] parseUntil Parse the packet until it reaches this protocol. The default value is ::UnknownProtocol which means don't take this parameter into account
+		 * @param[in] parseUntilLayer Parse the packet until certain layer in OSI model. The default value is ::OsiModelLayerUnknown which means don't take this parameter into account
+		 */
+		static void parseLayers(Packet* pkt, ProtocolType parseUntil = UnknownProtocol, OsiModelLayer parseUntilLayer = OsiModelLayerUnknown);

```

Comment:
I'm not sure I understand this API. Why putting both `parseUntil` and `parseUntilLayer` in the same method? why not separating them to two different methods? I think it'll be more clear that way

Reply:
1. There are `setRawPacket` method and a constructor in `Packet` that take both of these parameters
2. Having two parameters we can parse until a certain OSI model layer and stop if an inappropriate protocol has been reached. The simplest example is:
```
parseLayers(pkt, ProtocolType::ICMP, OsiModelLayer::OsiModelTransportLayer)
```
This call allows to parse until `TCP/UDP` layer and stops if an `ICMP` protocol is present.

I can append two methods:
```
static void parseLayers(Packet* pkt, ProtocolType parseUntil);
static void parseLayers(Packet* pkt, OsiModelLayer parseUntilLayer);
```

Reply:
Yes, you're right. I forgot I added it myself. My bad 😄